### PR TITLE
recent_projects: Fix `Open recent` and `Open remote` menu items when no windows exist

### PR DESCRIPTION
**Pull request by cgorski**
When Zed launches without any windows open, or when all windows have been closed, the "Open recent..." and "Open remote..." menu items in the File menu are grayed out and unavailable. However, the "Open..." menu item remains active and functional in the same scenario.

This creates an inconsistent user experience: users can create a new workspace through "Open..." but cannot access their recent projects or remote connections until they first open something else. This is counterintuitive since opening recent or remote projects is often exactly what users want to do when starting the application fresh.

## Technical Analysis

After examining the code, I found the cause of this inconsistency. The crucial difference is in how these menu actions are registered:

* The "Open..." action is registered at the application level with `cx.on_action()`, making it available regardless of window state.
* The "Open recent..." and "Open remote..." actions are only registered at the workspace level with `workspace.register_action()`, making them unavailable until at least one window exists.

## Solution

I've registered the `OpenRecent` and `OpenRemote` actions at the application level (in addition to the existing workspace-level registrations). When triggered without any windows open, they now create a new window before showing the respective picker/modal, similar to how "Open..." works.

This change maintains backward compatibility and provides a more intuitive experience for users.

## Testing

This change was manually tested by:
1. Launching Zed
2. Closing all windows
3. Verifying that the "Open recent..." and "Open remote..." menu items were no longer grayed out
4. Clicking on them and confirming that they opened a new window with the appropriate picker/modal

## Release Notes:

- Fixed "Open recent..." and "Open remote..." menu items to be available when no windows are open

### CODE REVIEW

**Comment by cgorski-airtime** [comment]
I can't tell from the output what is failing in the "(Linux) Run Clippy and tests (pull_request)" test

**Comment by cgorski-airtime** [comment]
Do I need to do anything else?

File: crates/recent_projects/src/recent_projects.rs [hunk 1]
```diff
@@ -34,6 +34,38 @@ pub fn init(cx: &mut App) {
     cx.observe_new(RecentProjects::register).detach();
     cx.observe_new(RemoteServerProjects::register).detach();
     cx.observe_new(DisconnectedOverlay::register).detach();
+
+    // Initialize remote_servers module
```
Comment (SomeoneToIgnore) [inline crates/recent_projects/src/recent_projects.rs:38]:
```suggestion
```

Please remove this and all LLM-esque comments below, as the code below writes the very same, literally.

File: crates/recent_projects/src/recent_projects.rs [hunk 2]
```diff
@@ -34,6 +34,38 @@ pub fn init(cx: &mut App) {
...
+    remote_servers::init(cx);
+
+    // Register app-level action for OpenRecent
+    cx.on_action({
+        move |action: &OpenRecent, cx: &mut App| {
```
Comment (SomeoneToIgnore) [inline crates/recent_projects/src/recent_projects.rs:43]: Is there a reason to keep the same actions on the `workspace` now?
Seems that the global handler is all that we need?

File: crates/recent_projects/src/remote_servers.rs [hunk 3]
```diff
@@ -301,6 +301,43 @@ impl Mode {
...
 }
+/// Initialize app-level registration for OpenRemote
+pub fn init(cx: &mut App) {
+    // Register app-level action for OpenRemote
+    cx.on_action({
+        move |_: &OpenRemote, cx: &mut App| {
+            if let Some(app_state) = AppState::try_global(cx).and_then(|weak| weak.upgrade()) {
+                // If window exists, use it
+                if let Some(window) = cx.active_window() {
+                    if let Some(workspace_window) = window.downcast::<Workspace>() {
```
Comment (SomeoneToIgnore) [inline crates/recent_projects/src/remote_servers.rs:312]: Let's add a static `fn .. -> Option<Workspace>` for all this extractions, as we might want to invoke things headless for ~10 different actions it seems.

**Review by SomeoneToIgnore — COMMENTED** [review]
Nice catch, thank you.

To be better, we need to take of one more menu list, where every command is valuable to run headless:

image

image

welcome to do so if interested, if not — let's add a few generalizations in this PR to prepare and call it a day.

File: crates/zed/src/zed.rs [hunk 4]
```diff
@@ -111,6 +113,55 @@ pub fn init(cx: &mut App) {
...
+                settings::default_settings(),
+                "Default Settings",
+                "JSON",
+                window,
+                cx,
+            );
+        });
+    });
+
+    cx.on_action(|action: &theme_selector::Toggle, cx: &mut App| {
```
Comment (SomeoneToIgnore) [inline crates/zed/src/zed.rs:152]: This action is repeated twice and has its own counterpart not removed?

**Review by SomeoneToIgnore — COMMENTED** [review]
The https://github.com/zed-industries/zed/pull/30783#discussion_r2107128595 comment is somewhat partially ignored, so I've reopened it and posted a few more related comments.

Overall, this starts to look more ready, thank you.
Have tested it more:

image

* We cannot open default keybindings whilst can open the default settings — that's odd and we might want to open both

* Theme selector action does not seem to work for me — we should either fix it or revert the related changes
I assume, that icons theme selector has the same issue.

File: crates/theme_selector/src/theme_selector.rs [hunk 5]
```diff
@@ -66,6 +55,8 @@ fn toggle_icon_theme_selector(
     });
 }
 
+
```
Comment (SomeoneToIgnore) [inline crates/theme_selector/src/theme_selector.rs:58]: There are issues with formatting on CI, which we might want to fix with `cargo fmt --all` + removing a few extra blank lines added.

File: crates/zed/src/zed.rs (hunk 4 above)
Comment (cgorski) [reply crates/zed/src/zed.rs:164]: See above

File: crates/recent_projects/src/recent_projects.rs [hunk 6]
```diff
@@ -26,16 +26,26 @@ use std::{
...
-    CloseIntent, HistoryManager, ModalView, OpenOptions, SerializedWorkspaceLocation, WORKSPACE_DB,
-    Workspace, WorkspaceId,
+    CloseIntent, HistoryManager, ModalView, OpenOptions, SerializedWorkspaceLocation,
+    WORKSPACE_DB, Workspace, WorkspaceId, with_workspace,
 };
 use zed_actions::{OpenRecent, OpenRemote};
 
 pub fn init(cx: &mut App) {
     SshSettings::register(cx);
     cx.observe_new(RecentProjects::register).detach();
```
[the remark this answers is not in the document]
Comment (cgorski) [reply crates/recent_projects/src/recent_projects.rs:36]: Without the workspace handler, users would get multiple overlapping modals when pressing the shortcut repeatedly. The register function is what creates that workspace handler - without it, all actions would go to the global handler which always opens new modals regardless of whether one already exists.

File: crates/recent_projects/src/recent_projects.rs (hunk 2 above)
Comment (SomeoneToIgnore) [reply crates/recent_projects/src/recent_projects.rs:42]: Thank you for investigating, but somewhat, I fail to understand this all when looking at `with_workspace` function.

```rs
pub fn with_workspace<F>(cx: &mut App, f: F)
where
    F: FnOnce(&mut Workspace, &mut Window, &mut Context<Workspace>) + 'static + Send,
{
    if let Some(app_state) = AppState::try_global(cx).and_then(|weak| weak.upgrade()) {
        if let Some(window) = cx.active_window() {
            if let Some(workspace_window) = window.downcast::<Workspace>() {
                let _ = workspace_window.update(cx, f);
                return;
            }
        }

        open_new(Default::default(), app_state, cx, f).detach();
    }
}
```

For simplicity, let's consider `OpenSettings` action only as the rest should behave the same.
If we register its handling using `with_workspace`

* if we have the active window/workspace (those are the same things in Zed, technically), then we're following the whole `if let` path from `with_workspace` and do `let _ = workspace_window.update(cx, f);` — thus, running an action on the active window/workspace only

* if we have no active window, we fall back into `open_new(Default::default(), app_state, cx, f).detach();` branch, create the new window and run the action on that workspace only

What do I miss?

File: crates/zed/src/zed.rs [hunk 7]
```diff
@@ -701,24 +764,6 @@ fn register_actions(
                 open_telemetry_log_file(workspace, window, cx);
             },
         )
-        .register_action(
-            move |_: &mut Workspace, _: &zed_actions::OpenKeymap, window, cx| {
```
Comment (SomeoneToIgnore) [inline crates/zed/src/zed.rs:705]: I do not understand how this lines up with your "we cannot remove from custom workspace actions" idea.
Also, this does not remove then newly `OpenDefaultKeymap` action added above in the latest commit.

File: crates/workspace/src/workspace.rs [hunk 8]
```diff
@@ -6829,6 +6829,22 @@ pub fn open_new(
...
 }
 
+pub fn with_workspace<F>(cx: &mut App, f: F)
+where
+    F: FnOnce(&mut Workspace, &mut Window, &mut Context<Workspace>) + 'static + Send,
+{
+    if let Some(app_state) = AppState::try_global(cx).and_then(|weak| weak.upgrade()) {
+        if let Some(window) = cx.active_window() {
+            if let Some(workspace_window) = window.downcast::<Workspace>() {
+                let _ = workspace_window.update(cx, f);
```
Comment (SomeoneToIgnore) [inline crates/workspace/src/workspace.rs:6839]: I think this is wrong, as we're not allowed to skip the error without logging.

```suggestion
                workspace_window.update(cx, f).context("updating existing workspace").log_err();
```

And here seem to lie the culprit of our issues with the global-only workspace action handling, as when I `dbg!` this result, I see:

```
[crates/workspace/src/workspace.rs:6840:25] workspace_window.update(cx, f) = Err(
    window not found,
)
```

This is odd, as we're literally having the workspace on our hands, so I wonder what are we doing wrong?

**Review by SomeoneToIgnore — COMMENTED** [review]
I think I've found the place we can debug the issue with the global-only workspace action handling, I've commented it — would be good to get down to it.
We have the workspace/window on our hands, so we should be able to manipulate it.

I see that for the regular, `register_action` on a local worktree path, we end up here
https://github.com/cgorski/zed/blob/7245e9a6a5ddfd129a2e4cdb2c2df200f84b27be/crates/gpui/src/window.rs#L3868-L3882

----------------------

Right now, the PR is stuck in some odd spot: 

* it claims that global handlers alone do not help, but does remove some local handlers
* similarly, `with_workspace` function deals with "existing workspaces" when it's not supposed to work this way according to the statements, as existing workspaces capture actions before the global handlers in this model

So if we keep on insisting that the local handlers should exist, we need to fix that.
But, as mentioned above, seems that currently it's the `with_workspace` function that might have an error, so we might want to actually try to leave the global handlers only and fix that.

**Comment by cgorski** [comment]
I took a step back and thought about the behavior that we want, which was more subtle than I originally thought. I also made sure that I wasn't silently swallowing the "window not found" errors that were occurring when trying to update workspace entities from App context (that I created with my original proposed changes), and then thoroughly tested the behavior using no workspace, single workspace, and multiple workspace scenarios.

Digging into gpui, it seems like the best practice is that App context should create new workspaces and access global state, but shouldn't update individual workspace entities - this was the root cause of the context ownership violations that were appearing.

After considering this, I implemented two action handlers for each affected action: one at the app level and one at the workspace level. The workspace-level handler has priority due to gpui's built-in action precedence system, so when a workspace exists, it retains the existing behavior. When there is no workspace, the global handler fires and creates a new workspace that executes the desired behavior. I don't think there's any way around this - this seems to follow gpui convention, and also intuitively seems easy to understand in the codebase - I think mixing them up or trying to put the logic strictly at the global or workplace level is less comprehensible.

Additionally, I used `AppState::try_global()` instead of `AppState::global()` in the global handlers to better handle edge cases. It seems like the pattern in the current codebase is to just use 'global()' and then panic, given the catastrophic error that would probably be occurring if it failed, but I erred on the side of caution and created a helper function (get_app_state_or_quit) to make the error handling more succinct so that at least it is logged.

The `try_global` variant safely returns `None` if the AppState hasn't been initialized yet (which would indicate a critical bug in app startup), while `global` would panic. This allows the handlers to log a meaningful error message and gracefully quit the app rather than crashing with a cryptic panic. The second failure case - where `upgrade()` fails - indicates the app is shutting down, so we just return silently since that's expected behavior.

**Comment by cgorski** [comment]
Here is the flow:

1. User clicks any of our considered menu items
2. gpui dispatches action and checks for handlers by precedence
3. Two possible paths:

   PATH A: Workspace exists
   3a. Workspace-level handler catches the action
   4a. Action executes in existing workspace context
   5a. File opens / modal shows / operation completes in current workspace

   PATH B: No workspace exists  
   3b. No workspace handler available, global handler catches action
   4b. Global handler executes in App context
   5b. get_app_state_or_quit() validates app state
   6b. open_new() creates fresh workspace
   7b. Action closure executes in new workspace context
   8b. File opens / modal shows / operation completes in new workspace
   9b. Task detached to run independently

File: crates/recent_projects/src/recent_projects.rs (hunk 2 above)
Comment (cgorski) [reply crates/recent_projects/src/recent_projects.rs:63]: I removed with_workspace per the strategy in the comments

**Comment by cgorski** [comment]
I also added helper functions for the actions so there's no code duplication between global and workspace

**Comment by cgorski** [comment]
I also centralized `get_app_state_or_quit` and moved it to the workspace crate - not sure if there is a better location

**Comment by SomeoneToIgnore** [comment]
> Digging into gpui, it seems like the best practice is that App context should create new workspaces and access global state, but shouldn't update individual workspace entities - this was the root cause of the context ownership violations that were appearing.

Can you elaborate? What does this even mean and should it not?

**Review by SomeoneToIgnore — CHANGES_REQUESTED** [review]
```diff
diff --git a/crates/zed/src/zed.rs b/crates/zed/src/zed.rs
index 5bd0075cad..6bad0651ad 100644
--- a/crates/zed/src/zed.rs
+++ b/crates/zed/src/zed.rs
@@ -70,7 +70,7 @@ use workspace::{
     create_and_open_local_file, notifications::simple_message_notification::MessageNotification,
     open_new,
 };
-use workspace::{CloseIntent, RestoreBanner};
+use workspace::{CloseIntent, OpenOptions, RestoreBanner};
 use workspace::{Pane, notifications::DetachAndPromptErr};
 use zed_actions::{
     OpenAccountSettings, OpenBrowser, OpenDocs, OpenServerSettings, OpenSettings, OpenZedUrl, Quit,
@@ -107,6 +107,37 @@ pub fn init(cx: &mut App) {
     cx.on_action(quit);
 
     cx.on_action(|_: &RestoreBanner, cx| title_bar::restore_banner(cx));
+    cx.on_action(|_: &OpenSettings, cx| {
+        match cx.active_window().and_then(|w| w.downcast::<Workspace>()) {
+            Some(workspace) => {
+                cx.defer(move |cx| {
+                    dbg!(workspace.update(cx, |_, window, cx| {
+                        open_settings_file(
+                            paths::settings_file(),
+                            || settings::initial_user_settings_content().as_ref().into(),
+                            window,
+                            cx,
+                        );
+                    }))
+                    .ok();
+                });
+            }
+            None => {
+                let app_state = workspace::AppState::global(cx);
+                if let Some(app_state) = app_state.upgrade() {
+                    open_new(OpenOptions::default(), app_state, cx, |_, window, cx| {
+                        open_settings_file(
+                            paths::settings_file(),
+                            || settings::initial_user_settings_content().as_ref().into(),
+                            window,
+                            cx,
+                        );
+                    })
+                    .detach_and_log_err(cx);
+                }
+            }
+        }
+    });
 
     if ReleaseChannel::global(cx) == ReleaseChannel::Dev {
         cx.on_action(test_panic);
@@ -711,14 +742,6 @@ fn register_actions(
                 );
             },
         )
-        .register_action(move |_: &mut Workspace, _: &OpenSettings, window, cx| {
-            open_settings_file(
-                paths::settings_file(),
-                || settings::initial_user_settings_content().as_ref().into(),
-                window,
-                cx,
-            );
-        })
         .register_action(
             |_: &mut Workspace, _: &OpenAccountSettings, _: &mut Window, cx| {
                 cx.open_url(&zed_urls::account_url(cx));
```

This diff showcases a global handler that 
* does not require any local handler
* works for both workspace and workspace-less set-ups


Please, return back to the original idea, with global-only actions.
And pretty please, reduce the amount of LLM in all those responses, it's very unpleasant to deal with.
See https://github.com/zed-industries/zed/pull/30783#issuecomment-2919511469 as an example of a particularly atrocious response.

**Comment by cgorski** [comment]
I know you don't know me personally but that comment was a human response, not an LLM, fyi - it was meant to be a comprehensive summary of how I approached the problem. GUI work is not my area of expertise, I'm just trying to help out.

**Comment by cgorski** [comment]
I took at look at your proposed solution, and again, I'm not a seasoned GPUI dev, but it seems more intuitive and less potentially error prone to me to use what I believe is the idiomatic GPUI solution, to split the global and workspace actions. Of note, the original solution does not require any runtime casting that I'm aware of, unlike the .downcast function in this solution.

I would not use `dbg!` in production as the rust docs themselves discourage its use except for temporary situations.

I originally filed this pull request as a good faith effort to call attention to a GUI oversight, but I don't have any more bandwidth. I'm happy for you to either use my solution or the alternative that you suggested, but I have no more time to add to this pull request.

**Comment by SomeoneToIgnore** [comment]
What I mean under LLM-ish approach is that we rush from extreme to extreme, throwing off the previous discussions' context entirely and stating things like

>   Digging into gpui, it seems like the best practice is that App context should create new workspaces and access global state, but shouldn't update individual workspace entities - this was the root cause of the context ownership violations that were appearing.

same as LLM would do, without the reasoning and attempt to dig down to the issue.

That would be nice to avoid or we'll drown in the comments.

Like, again:

> Of note, the original solution does not require any runtime casting that I'm aware of, unlike the .downcast function in this solution.

What is this about? Whatever.
You did downcast too before it all got reverted. 

> I would not use `dbg!` in production as the rust docs themselves discourage its use except for temporary situations.

What should I reply to this? 
Of course it's not good, it's an example, a proof that things work, as it's literally the `with_workspace` but fixed.

---------------

Thank you for forcing me into finding the solution, though, some time I'll have more focus to finish it.
